experiments/compositions/composition/tests/kind/cluster.go

```diff
@@ -0,0 +1,235 @@
+	var cluster KindClusterReader
+	found := false
+	for !found {
+		kindClusterSet.available.Range(func(k, v any) bool {
+			v, loaded := kindClusterSet.available.LoadAndDelete(k)
+			if !loaded {
+				return true
+			}
+			found = true
+			cluster = v.(KindClusterReader)
+			return false
+		})
+
+		if found {
+			continue
+		}
+
+		t.Logf("\nWaiting for a cluster to become available...")
+		time.Sleep(5 * time.Second)
+	}
+
+	t.Logf("Reserved cluster %s", cluster.Name())
+	return cluster
+}
+
+func ReleaseCluster(t *testing.T, cluster KindClusterReader) {
+	kindClusterSet.available.Store(cluster.Name(), cluster)
+	t.Logf("Released cluster %s", cluster.Name())
+}
+
+func verifyKindIsInstalled() error {
+	_, err := exec.LookPath("kind")
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// NewKindCluster - return a cluster setup object
+func NewKindCluster(name string) (KindCluster, error) {
+	err := verifyKindIsInstalled()
+	if err != nil {
+		return nil, err
+	}
+	cluster := &kindCluster{
+		name: name,
+	}
+	return cluster, nil
+}
+
+// Wait for all clusters to become ready
+func (c *kindCluster) Create() error {
+	err := c.Delete()
+	if err != nil {
+		return err
+	}
+	clusterConfig, err := c.kindClusterDefinition()
+	if err != nil {
+		return err
+	}
+	defer os.Remove(clusterConfig)
+
+	c.config, err = c.createCluster(clusterConfig)
+
+	kindClusterSet.available.Store(c.Name(), c)
+	return nil
+}
+
+// Config return rest.Config
+func (c *kindCluster) Config() *rest.Config {
+	return c.config
+}
+
+// Name return name
+func (c *kindCluster) Name() string {
+	return c.name
+}
+
+func (c *kindCluster) String() string {
+	return c.name
+}
+
+// LoadImage loads a docker image into the cluster
+func (c *kindCluster) LoadImage(image string) error {
+	err := exec.Command("kind", "load", "docker-image", image, "--name", c.name).Run()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Delete deletes all clusters
+func (c *kindCluster) Delete() error {
+	kindClusterSet.available.Delete(c.Name())
+	err := exec.Command("kind", "delete", "cluster", "--name", c.name).Run()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Exists checks if the cluster exists
+func (c *kindCluster) Exists() (bool, error) {
+	output, err := exec.Command("kind", "get", "clusters").CombinedOutput()
+	if err != nil {
+		return false, err
+	}
+
+	for _, cluster := range strings.Split(string(output), "\n") {
+		if cluster == c.name {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (c *kindCluster) kindClusterDefinition() (string, error) {
+	ipAddress, err := c.getHostIPAddress()
+	if err != nil {
+		return "", err
+	}
+	clusterConfigFile, err := os.CreateTemp("", "kind-cluster.yaml")
+	if err != nil {
+		return "", err
+	}
+	defer clusterConfigFile.Close()
+	kindClusterConfig := `kind: Cluster
+apiVersion: kind.x-k8s.io/v1alpha4
+networking:
+  # Allow connections to the API Sever with the CloudTop IP address
```

Comment:
Nit: typo in server, and can we make "CloudTop" more universally relevant?  Maybe "host IP address".  Also, we might not need (or want) this here

Reply:
fixed.